lib/dry/schema/macros/filled.rb

```diff
@@ -20,11 +27,35 @@ def call(*predicates, **opts, &block)
           end
 
           if opts[:type_spec]
-            value(predicates[0], :filled?, *predicates[1..predicates.size - 1], **opts, &block)
+            if filter?
+              value(*predicates, **opts, &block)
+            else
+              value(predicates[0], :filled?, *predicates[1..predicates.size - 1], **opts, &block)
+            end
           else
             value(:filled?, *predicates, **opts, &block)
           end
         end
+
+        # @api private
+        def filter?
+          !primitives.include?(NilClass) && processor_config.filter_empty_string
+        end
```

Comment:
@flash-gordon @timriley what this means is: "if `nil` is NOT an allowed value, and empty strings should be filtered out, then we want to apply `filter(:filled?)`"

Reply:
I wonder if we can make this clearer (and less needing explanatory comment like the one you left here) by making some of the names more self-descriptive?

e.g. `def handle_via_filter?` for this method, and `expected_primitives` instead of `primitives`?